packages/playwright-test/src/runner.ts

```diff
@@ -414,24 +487,32 @@ export class Runner {
 
     // 14. Run tests.
     try {
-      const sigintWatcher = new SigIntWatcher();
+      let sigintWatcher;
 
       let hasWorkerErrors = false;
-      const dispatcher = new Dispatcher(this._loader, testGroups, this._reporter);
-      await Promise.race([dispatcher.run(), sigintWatcher.promise()]);
-      if (!sigintWatcher.hadSignal()) {
-        // We know for sure there was no Ctrl+C, so we remove custom SIGINT handler
-        // as soon as we can.
-        sigintWatcher.disarm();
+      for (const testGroups of concurrentTestGroups) {
+        const dispatcher = new Dispatcher(this._loader, testGroups, this._reporter);
+        sigintWatcher = new SigIntWatcher();
+        await Promise.race([dispatcher.run(), sigintWatcher.promise()]);
+        if (!sigintWatcher.hadSignal()) {
+          // We know for sure there was no Ctrl+C, so we remove custom SIGINT handler
+          // as soon as we can.
+          sigintWatcher.disarm();
+        }
+        await dispatcher.stop();
```

Comment:
You miss Ctrl+C here.

Reply:
I would keep single sigint watcher around, would fail out from the look if sigintWatcher.hadSignal. If Ctrl+C is pressed during dispatcher.stop it'll self-disarm and keep hanging. It'll complete stop and break the look gracefully.